### DISCUSSION TITLE
Import modules from decentralized networks

### DISCUSSION THREAD

**Discussion by alexgleason** [denoland/deno#22779] [Ideas]
Sounds crazy, but I would like to be able to do this:

```ts
import { Machina } from 'nostr:nevent1qqsvnhgrwkwyx96zqauqjzk2wwz25a39rx75h8d68r4wlhdyk9s849spzemhxue69uhhyetvv9ujumn0wd68ytnzv9hxgq3qjky53pl5p4v73qr7dr2l76ww895tgr5fznmt8aqwhy6ay39ssp6q3ls4xh';
```

You can check the contents of it here: https://njump.me/nevent1qqsvnhgrwkwyx96zqauqjzk2wwz25a39rx75h8d68r4wlhdyk9s849spzemhxue69uhhyetvv9ujumn0wd68ytnzv9hxgq3qjky53pl5p4v73qr7dr2l76ww895tgr5fznmt8aqwhy6ay39ssp6q3ls4xh

Nostr stores immutable events on a set of relays, making it suitable for importing ES modules.

Using import maps, it's actually already possible to import code from Nostr using an HTTPS gateway!

deno.json (adding my Nostr HTTPS gateway):

```diff
  "imports": {
+    "@nostr/": "https://77484d49eac4.ngrok.app/e/"
  },
```

code:

```ts
import { Machina } from '@nostr/nevent1qqsvnhgrwkwyx96zqauqjzk2wwz25a39rx75h8d68r4wlhdyk9s849spzemhxue69uhhyetvv9ujumn0wd68ytnzv9hxgq3qjky53pl5p4v73qr7dr2l76ww895tgr5fznmt8aqwhy6ay39ssp6q3ls4xh';
```

This works, and it's absolutely amazing!

This is possible thanks to [hostr](https://github.com/studiokaiji/nostr-webhost).

I just have one small wish - which is that import maps could let me specify a custom protocol. This **doesn't work**:

```json
  "imports": {
    "nostr:": "https://77484d49eac4.ngrok.app/e/"
  },
```

I wish import maps either allowed this syntax, or a new `protocols` field could be added to deno.json to enable this functionality.

It's a small complaint. Deno is amazing.

**Comment by vitorpamplona** [comment]
This is awesome!

**Comment by alexgleason** [comment]
I am also very inspired by [JSR](https://jsr.io/), a small step towards decentralizing NPM, that I could imagine being entirely decentralized and content-addressable, with tools to publish packages on Nostr relays or similar. I imagine it could use a Nostr relay for its entire database - for storing project metadata and source code files. Just a thought.
